views/utils.py

```diff
@@ -225,30 +225,53 @@ def create_graph_for_interfaces(datasources, start=None, end=None, **kwargs):
 
 def create_links_dict(links):
     ifces = {}
+    ifce_pairs = {}
+    # walk through links to get all interfaces
     for link in links:
         try:
-            if ifces.get(
-                '%s_%s' % (
-                    link.local_ifce.node.name, link.remote_ifce.node.name
-                )
-            ):
-                name = '%s_%s' % (link.local_ifce.node.name, link.remote_ifce.node.name)
-                prev = ifces.get(name)
-                prev.append(link.local_ifce.as_dict())
-                ifces.update({
-                    '%s_%s' % (
-                        link.local_ifce.node.name, link.remote_ifce.node.name
-                    ): prev
-                })
+            ifces.setdefault(link.local_ifce.pk, link.local_ifce)
+            ifces.setdefault(link.remote_ifce.pk, link.remote_ifce)
+        except AttributeError:
+            continue
+
+    # fetch all DataSource objects for our set of interfaces,
+    # pre-loading related RrdFile objects
+    dss = DataSource.objects.\
+        filter(object_id__in=ifces.keys(),
+               content_type__model='ifce',
+               graph__type='traffic').\
+        select_related('rrdfile')
+    # force QuerySet to be evaluated, returning a list of tuples
+    # indexed by ifce
+    dss = [(x.object_id, x) for x in dss]
```

Comment:
`select_related()` determines additional models/fields Django ORM will fetch for this QuerySet when it is evaluated. So this just adds `RrdFile` fields to the query that is performed once, in this list comprehension, avoiding the need for a subsequent db query in the next iteration.